tests/filtered_relation/tests.py

```diff
@@ -0,0 +1,383 @@
+        self.assertSequenceEqual(qs1.difference(qs2), [self.author1])
+
+    def test_select_for_update(self):
+        self.assertSequenceEqual(
+            Author.objects.annotate(
+                book_jane=FilteredRelation('book', Q(book__title__iexact='the book by jane a')),
+            ).filter(book_jane__isnull=False).select_for_update(),
+            [self.author2]
+        )
+
+    def test_defer(self):
+        # One query for the list and one query for the deferred title.
+        with self.assertNumQueries(2):
+            self.assertQuerysetEqual(
+                Author.objects.annotate(
+                    book_alice=FilteredRelation('book', Q(book__title__iexact='poem by alice')),
+                ).filter(book_alice__isnull=False).select_related('book_alice').defer('book_alice__title'),
+                ['Poem by Alice'], lambda author: author.book_alice.title
+            )
+
+    def test_only_not_supported(self):
+        msg = 'only() is not supported with FilteredRelation.'
+        with self.assertRaisesMessage(ValueError, msg):
+            Author.objects.annotate(
+                book_alice=FilteredRelation('book', Q(book__title__iexact='poem by alice')),
+            ).filter(book_alice__isnull=False).select_related('book_alice').only('book_alice__state')
+
+    def test_as_subquery(self):
+        inner_qs = Author.objects.annotate(
+            book_alice=FilteredRelation('book', Q(book__title__iexact='poem by alice')),
+        ).filter(book_alice__isnull=False)
+        qs = Author.objects.filter(id__in=inner_qs)
+        self.assertSequenceEqual(qs, [self.author1])
+
+    def test_with_foreign_key_error(self):
+        msg = (
+            "Filtered relation 'alice_favorite_books' cannot operate on "
+            "foreign key 'author__favorite_books__author'."
+        )
+        with self.assertRaisesMessage(ValueError, msg):
+            list(
+                Book.objects.annotate(
+                    alice_favorite_books=FilteredRelation(
+                        'author__favorite_books',
+                        Q(author__favorite_books__author=self.author1),
+                    )
+                ).filter(alice_favorite_books__title__icontains='poem')
+            )
+
+    def test_with_foreign_key_on_condition_error(self):
+        msg = (
+            "Filtered relation 'book_edited_by_b' cannot operate on foreign "
+            "key 'book__editor__name__icontains'."
+        )
+        with self.assertRaisesMessage(ValueError, msg):
+            list(Author.objects.annotate(
+                book_edited_by_b=FilteredRelation('book', Q(book__editor__name__icontains='b')),
+            ).filter(book_edited_by_b__isnull=False))
+
+    def test_with_empty_relation_name_error(self):
+        with self.assertRaisesMessage(ValueError, 'relation_name cannot be empty.'):
+            FilteredRelation('', Q(blank=''))
+
+    def test_filered_relation_with_condition_as_expression_error(self):
```

Comment:
This limitation is not documented. Are you fine with it ? I just want to double check it has not been removed by mistake.

Reply:
The documentation for the `condition` attribute says, "A Q object to control the filtering.". That seems sufficient to me.